### ISSUE TITLE
CTC GPU support

### ISSUE THREAD

**Issue by Hibbert-pku** [tensorflow/tensorflow#2146]
Very happy to see CTC in  #tensorflow! #32
My question is:
Is someone working on CTC GPU implement internal? Or is it 'contribution welcomed'?
Baidu has a CTC GPU implement:

https://github.com/baidu-research/warp-ctc

**Comment by zheng-xq** [comment]
As Eugene and Vijay commented, it is currently on the contrib directory. And once it matures enough, it will be moved to the core directory. 

As always, contributions are welcome. For any significant code changes, it is optional to announce it first, so people would notice and avoid replicated work.

**Comment by Hibbert-pku** [comment]
Sorry to bother you again, can you point the ctc .cu.cc file, please? I can't find it.

**Comment by cjh9** [comment]
+1, CTC GPU support would be awesome for handwritten text recognition, speech recognition etc. May key applications here!! 🎉 😀

**Comment by Hibbert-pku** [comment]
I've merged tensorflow and baidu's warp-ctc in:
https://github.com/Hibbert-pku/tensorflow
maybe it's useful for you.
@cjh9

**Comment by igormq** [comment]
Anyone have news about it?

**Comment by wangkuiyi** [comment]
Any news?

@zheng-xq Can you please point where are Eugene and Vijay's comments, and the location of GPUCTC "on the contrib directory"?

**Comment by zheng-xq** [comment]
Adding @ebrevdo to comment on the status of CTC GPU implementation.

**Comment by ebrevdo** [comment]
Very much a "contributions welcome"; @Hibbert-pku has what looks like an implementation in their own github.  It includes some nvidia code which probably cannot be merged into the main tensorflow repository, and a lot of cleanup would have to happen before we could merge a pull request.

**Comment by lingz** [comment]
Maybe we could open this issue back up then, so it's visible?

**Comment by andykitchen** [comment]
I would also find a GPU implementation of CTC loss very helpful!

**Comment by rgobbel** [comment]
+1 for a GPU implementation of CTC loss.

**Comment by LinJM** [comment]
+ 1 for GPU CTC Loss @ @ # #

**Comment by standy66** [comment]
+1 It would be much appreciated to see GPU implementation of connectionist temporal classification loss in Tensorflow.

**Comment by rgobbel** [comment]
I managed to get the [**@MISingularity**](https://github.com/MISingularity/tensorflow) Warp CTC implementation working in post-0.10.0 TensorFlow. Not difficult, but as noted, it includes a bunch of proprietary code from NVIDIA. It also completely ignores the `preprocess_collapse_repeated` and `ctc_merge_repeated` flags, so if you need those, as I do, the only option is to pad the input with blanks, then remove the padding afterward. I decided to drop CTC and use a different sequence-to-sequence framework.

**Comment by jaredcasper** [comment]
Heads up that there is now a TensorFlow kernel that wraps warp-ctc in the master branch of the warp-ctc repository: https://github.com/baidu-research/warp-ctc

(note that it too does not support proprocess_collapse_repeated=true or ctc_merge_repeated=false)

**Comment by LinJM** [comment]
@jaredcasper where is the kernel? I can't find it.

**Comment by jaredcasper** [comment]
@LinJM it is in the tensorflow_binding directory (and only on the master branch) https://github.com/baidu-research/warp-ctc/tree/master/tensorflow_binding

**Comment by zakizhou** [comment]
Any progress?

**Comment by pannous** [comment]
can warp-ctc be merged into tensorflow?

**Comment by rgobbel** [comment]
As commented above, getting it to work wasn't hard, but I stopped working on it because of the dependencies on code with problematic licensing terms. If that's not an issue, it could be merged into the main TensorFlow sources pretty easily.

I started working on adding support for the preprocess_merge_repeated and ctc_merge_repeated flags, but that's tricky, involving quite a few changes to the kernel--small, but all over the place.

**Comment by matth** [comment]
Just to update this this ticket in case others come here looking for Warp CTC integration.

The Warp CTC maintainers have released a tensorflow binding which works transparently with `proprocess_collapse_repeated=False` and `ctc_merge_repeated=True`. 

I've tested this with Tensorflow 0.12.0 and got good results, though the speed improvement was marginal in my case.

https://github.com/baidu-research/warp-ctc/tree/master/tensorflow_binding

**Comment by 0x454447415244** [comment]
@matth I just tried it and I can't see an improvement in speed. I see the CPU utilization is the same in both cases (`tf.nn.ctc_loss` and `warpctc_tensorflow.ctc`).

**Comment by matth** [comment]
@0x454447415244 yes I see similar, not a noticeable improvement in speed and similar CPU usage. Not 100% sure what is happening, I' pretty sure it usese the GPU as I can see it getting compiled with CUDA support.

TBH I am not sure how much improvement I should expect to see, the warp ctc benchmarks do not compare against Tensorflow's implementation. 

Personally I am doing CTC with audio features and a mini-batch size of 64 takes 12GB of GPU RAM, so perhaps the CTC  loss calculation is a negligible part of the execution time anyway. I'll stick with using warp though as it's certainly not made my training slower or less accurate.

**Comment by ilovin** [comment]
I tried it and I can't see an improvement in CPU mode.
On the other hand, the performance in GPU mode(need four tensors as feed) seems unstable,
cannot get a good result from it.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by alextp** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
